staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go

```diff
@@ -206,3 +227,109 @@ func (t *vSphereCSITranslator) GetCSIPluginName() string {
+	zones := getTopologyValues(pv, zoneLabel)
+	if len(zones) > 0 {
+		replaceTopology(pv, zoneLabel, csiTopologyKeyZone)
+	} else {
+		// if nothing is in the NodeAffinity, try to fetch the topology from PV labels
+		if label, ok := pv.Labels[zoneLabel]; ok {
+			if len(label) > 0 {
+				addTopology(pv, csiTopologyKeyZone, []string{label})
+			}
+		}
+	}
+
+	// If region kubernetes topology exist, replace it to use csiTopologyKeyRegion
+	regions := getTopologyValues(pv, regionLabel)
+	if len(regions) > 0 {
+		replaceTopology(pv, regionLabel, csiTopologyKeyRegion)
+	} else {
+		// if nothing is in the NodeAffinity, try to fetch the topology from PV labels
+		if label, ok := pv.Labels[regionLabel]; ok {
+			if len(label) > 0 {
+				addTopology(pv, csiTopologyKeyRegion, []string{label})
+			}
+		}
+	}
+	return nil
+}
+
+// translateTopologyFromCSIToInTreevSphere converts CSI zone/region affinity rules to in-tree vSphere zone/region labels
+func translateTopologyFromCSIToInTreevSphere(pv *v1.PersistentVolume,
+	csiTopologyKeyZone string, csiTopologyKeyRegion string) error {
+	zoneLabel, regionLabel := getTopologyLabel(pv)
+
+	// Replace all CSI topology to Kubernetes Zone label
+	err := replaceTopology(pv, csiTopologyKeyZone, zoneLabel)
+	if err != nil {
+		return fmt.Errorf("failed to replace CSI topology to Kubernetes topology, error: %v", err)
+	}
+
+	// Replace all CSI topology to Kubernetes Region label
+	err = replaceTopology(pv, csiTopologyKeyRegion, regionLabel)
+	if err != nil {
+		return fmt.Errorf("failed to replace CSI topology to Kubernetes topology, error: %v", err)
+	}
+
+	zoneVals := getTopologyValues(pv, zoneLabel)
+	if len(zoneVals) > 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		_, zoneOK := pv.Labels[zoneLabel]
+		if !zoneOK {
+			pv.Labels[zoneLabel] = zoneVals[0]
+		}
+	}
+	regionVals := getTopologyValues(pv, regionLabel)
+	if len(regionVals) > 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		_, regionOK := pv.Labels[regionLabel]
+		if !regionOK {
+			pv.Labels[regionLabel] = regionVals[0]
+		}
+	}
+	return nil
+}
+
+// translateAllowedTopologiesForvSphereCSI translates allowed topologies within storage class
+// from legacy failure domain to vSphere CSI topology keys
+func translateAllowedTopologiesForvSphereCSI(terms []v1.TopologySelectorTerm, zonekey string, regionkey string) ([]v1.TopologySelectorTerm, error) {
+	if terms == nil {
+		return nil, nil
+	}
+
+	newTopologies := []v1.TopologySelectorTerm{}
+	for _, term := range terms {
+		newTerm := v1.TopologySelectorTerm{}
+		for _, exp := range term.MatchLabelExpressions {
+			var newExp v1.TopologySelectorLabelRequirement
+			if exp.Key == v1.LabelFailureDomainBetaZone || exp.Key == v1.LabelTopologyZone {
+				newExp = v1.TopologySelectorLabelRequirement{
+					Key:    zonekey,
+					Values: exp.Values,
+				}
+			} else if exp.Key == v1.LabelFailureDomainBetaRegion || exp.Key == v1.LabelTopologyRegion {
```

Comment:
SC Allowed topologies did not allow region labels right? Why is this needed?

Reply:
Yes, and the vSphere in-tree plugin does not read region while creating volume.
https://github.com/kubernetes/kubernetes/blob/06e107081e4e0182efb9daa0daf7f2ae12272704/pkg/volume/vsphere_volume/vsphere_volume.go#L370-L376

I am removing the translation of region then.